Update PSE 7 with new question flow (#63)
Ada-Developers-Academy/core-pses: 07/reshape-a-matrix.md

1. The last two bullet points of the description say that we should output a new matrix or raise a ValueError.
    - This tells me that the only expected return value of the function should be a new matrix if reshaping is possible.

2. The problem statement says the matrix we receive as an argument will be represented as a 2 dimensional array. In the examples, the outputs are also displayed as 2 dimensional arrays.
    - This means that the new matrix we create should also be a 2 dimensional array.

3. When discussing the output, it says we should raise a ValueError if the original matrix cannot be reshaped as requested.
    - I will need to do some check in my function to see if the number of items in the original matrix is equal to the number of items in the requested matrix before trying the reshape operation.

4. The problem statement says the new matrix needs to be filled with the elements of the old matrix in the same row-traversing order.
    - This means that the order of the elements in the new matrix matters. I will need to go row by row starting at the front of each list in the original matrix as I fill up the new matrix.

> 5. We can assume that the arguments to the function will be valid, so the input matrix will be a 2 dimensional list, and the `rows` and `columns` will not be negative. We have also been given the note "It can be assumed that all given `rows` and `columns` are positive numbers." 

Reviewer: 2 dimensional list just doesn't sit well with me. Like I'd be totally OK with 2 dimensional array, even though there would be no structural difference. In the grand scheme of things, this isn't a big deal, but here's an alternative phrasing to consider, which also helps the reader rule out jagged inputs.
```suggestion
5. We can assume that the arguments to the function will be valid, so the input matrix will be a nested list with a consistent number of `rows` and `columns`, and the `rows` and `columns` will not be negative. We have also been given the note "It can be assumed that all given `rows` and `columns` are positive numbers." 
```

<br>

In the box below, please submit:
1. A shareable link to your conversation in ChatGPT
    - [Documentation for creating a shareable link in ChatGPT](https://help.openai.com/en/articles/7925741-chatgpt-shared-links-faq)
2. Your reflections and summary of the discussion with ChatGPT

##### !end-question
##### !hint

**Troubleshooting**
- If you are having issues with the tool understanding the prompt, try formatting the problem statement or examples differently.
- If you’ve reformatted the information and are still not getting useful results, reach out in #study-hall and share what you are experiencing and the link to your chat so folks can take a look and help you troubleshoot!

<br>

**Summarizing the Review**
- Did the AI tool uncover anything about the observations you made that you hadn’t considered?
- Did the AI tool uncover anything about the questions you asked that you hadn’t considered?
- Did the AI tool suggest updates to the observations you made or questions you asked? 
    - If so, what updates and why?
- Did the AI tool suggest any new observations or questions?
    - If so, what? Why would they be useful?

##### !end-hint
##### !explanation 

> For an example of what a review response might look like, let’s say that we provided observations simliar to the example response from the "Explanation" section of the previous question to complete the review prompt. 

Reviewer: ```suggestion
For an example of what a review response might look like, let’s say that we provided observations similar to the example response from the "Explanation" section of the previous question to complete the review prompt. 
```

<br>

Depending on how our observations differ and exactly what ChatGPT shares, a reflection and summary might look like:

<br>

Chat link: `<url to your conversation>`

<br>

> My feedback from ChatGPT suggested I could strengthen my list by asking about whether matrices must always be rectangular or could be jagged which would help cover edge cases, define the boundaries of the problem space, and clarify assumptions.

Reviewer: What if following the bit about jagged inputs, we included something like
```suggestion
My feedback from ChatGPT suggested I could strengthen my list by asking about whether matrices must always be rectangular or could be jagged which would help cover edge cases, define the boundaries of the problem space, and clarify assumptions. I asked some follow up questions to better understand what a jagged array is. Since it wouldn't be a valid matrix, it sounds like I can ignore that possibility for this case.
```

One of many possible responses could look like:

1. The last two bullet points of the description say that we should output a new matrix or raise a ValueError.
    - This tells me that the only expected return value of the function should be a new matrix if reshaping is possible.

2. The problem statment says the matrix we receive as an argument will be represented as a 2 dimensional array. In the examples, the outputs are also displayed as 2 dimensional arrays.
    - This means that the new matrix we create should also be a 2 dimensional array.

3. When discussing the output, it says we should raise a ValueError if the original matrix cannot be reshaped as requested.
    - I will need to do some check in my function to see if the number of items in the original matrix is equal to the number of items in the requested matrix before trying the reshape operation.

4. The problem statment says the new matrix needs to be filled with the elements of the old matrix in the same row-traversing order.
    - This means that the order of the elements in the new matrix matters. I will need to go row by row starting at the front of each list in the original matrix as I fill up the new matrix.

> 5. We can assume that the arguments to the function will be valid, so the input matrix will be a 2 dimensional list, and the `rows` and `columns` will not be negative, but are there any special considerations if the input matrix is empty?

Author: I'm more inclined to change the observation to note that we do not need to consider the empty case at all (I must have forgotten about the note at the end of the problem description while writing this.)

This is what I have it changed to locally:

5. We can assume that the arguments to the function will be valid, so the input matrix will be a 2 dimensional list, and the `rows` and `columns` will not be negative. We have also been given the note "It can be assumed that all given `rows` and `columns` are positive numbers." 
    - This means that the implementation does not need to worry about a case where the input matrix is empty or how an empty matrix might be represented in this scenario, as it is unclear what the definition of an empty matrix could be for this problem. (For example `[]` reports 0 rows, but would error if we try to read the columns and `[[]]` reports 0 columns, but 1 row).

<br>

In the box below, please submit:
1. A shareable link to your conversation in ChatGPT
    - [Documentation for creating a shareable link in ChatGPT](https://help.openai.com/en/articles/7925741-chatgpt-shared-links-faq)
2. Your reflections and summary of the discussion with ChatGPT

##### !end-question
##### !hint

**Troubleshooting**
- If you are having issues with the tool understanding the prompt, try formatting the problem statement or examples differently.
- If you’ve reformatted the information and are still not getting useful results, reach out in #study-hall and share what you are experiencing and the link to your chat so folks can take a look and help you troubleshoot!

<br>

**Summarizing the Review**
- Did the AI tool uncover anything about the observations you made that you hadn’t considered?
- Did the AI tool uncover anything about the questions you asked that you hadn’t considered?
- Did the AI tool suggest updates to the observations you made or questions you asked? 
    - If so, what updates and why?
- Did the AI tool suggest any new observations or questions?
    - If so, what? Why would they be useful?

##### !end-hint
##### !explanation 

> For an example of what a review response might look like, let’s say that we used the example response from the "Explanation" section of the previous question to complete the review prompt. 

Author: It would be nice to replicate this to other PSEs, but I'm good with updating this here and adjusting the earlier PSEs when we have future tickets to batch that with since I don't see it being very impactful to students as a change, more nice for consistency and a little distancing from the observations & questions.

- The objective is to create a roadmap that we can use to keep ourselves oriented towards our goal
- It is okay to leave some of the finer details to be worked out in the implementation itself!

As you write your steps, keep the following guidelines in mind:
* We want to think about a general approach rather than what the code would look like line-by-line. 
* It may be helpful to break up the problem/algorithm into smaller subproblems/algorithms. 
    * For example: 1. Handle edge cases, 2. Perform the computation/solve the problem/etc.
* The steps should be a description as if you were talking out the problem with another person and should be agnostic of any particular language. 
    * As such, they should not include code syntax in the description.

What's important at this stage is to think through and outline the implementation before writing code.

##### !end-question
##### !placeholder

Write the logical steps here.

##### !end-placeholder
### !explanation

Example Steps:

1. Check if the reshape can be done:

>     - See if the number of elements in the original matrix (number of rows multiplied by the number of columns) matches the number of elements in the desired output (the argument `rows` multipled by the argument `columns`).

Reviewer: ```suggestion
    - See if the number of elements in the original matrix (number of rows multiplied by the number of columns) matches the number of elements in the desired output (the argument `rows` multiplied by the argument `columns`).
```
